api/bucketsd/service.go

```diff
@@ -146,6 +165,30 @@ func (s *Service) List(ctx context.Context, _ *pb.ListRequest) (*pb.ListResponse
 	return &pb.ListResponse{Roots: roots}, nil
 }
 
+func (s *Service) republish(session string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), listKeysTimeout)
+	defer cancel()
+	start := time.Now()
+	keys, err := s.Collections.IPNSKeys.List(ctx)
+	if err != nil {
+		return err
+	}
+	for _, key := range keys {
+		dctx := common.NewThreadIDContext(common.NewSessionContext(ctx, session), key.ThreadID)
+		pth, err := s.ListPath(dctx, &pb.ListPathRequest{
+			Key:  key.Cid,
+			Path: "/",
+		})
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		s.IPNSManager.Publish(path.New(pth.Item.Cid), key.Cid)
+	}
+	log.Infof("republishing finished %d records => %v", len(keys), time.Since(start))
+	return nil
```

Comment:
crux of it above

Reply:
Hmm, yeah it's a weird thing to do because the service doesn't really "own" the threads. There won't be a session/api key to add to the context since this is an internal operation (not user initiated). 

Rather than using `ListPath`, you can use `s.Buckets.List`:
```
list, err := s.Buckets.List(ctx, threadID, &db.Query{}, &tdb.Bucket{}, tdb.WithToken(dbToken))
if err != nil {
	return nil, err
}
bucks := list.([]*tdb.Bucket)
// loop over buckets
```

We're not providing a thread token here. So, you may be blocked from reading that thread... if you're not, that's only cause there's still some ACL related todos in `go-threads`. 

Another option is to modify the IPNS record mongo collection to store the most recent successfully published path. Then this fix will only work after a bucket it pushed to, which is probably fine. That would prevent future headaches with thread tokens as well.

Reply:
yep, cool. was trying to avoid the model change. but looking back now it seems so much cleaner. coming up